lib/cn/kblock_builder.c

```diff
@@ -1287,6 +1290,131 @@ kbb_set_merge_stats(struct kblock_builder *bld, struct cn_merge_stats *stats)
     bld->mstats = stats;
 }
 
+static merr_t
+kblock_copy_range(
+    struct kblock_desc   *kbd,
+    const struct key_obj *start, /* exclusive */
+    const struct key_obj *end,   /* inclusive */
+    uint64_t             *kb_out)
+{
+    struct wbti *wbti = NULL;
+    struct kblock_builder *kbb;
+    struct key_obj cur = { 0 };
+    const void *kmd;
+    merr_t err;
+    uint64_t tot_keys = 0;
+
+    /* start == NULL implies a copy from the first key
+     * end == NULL implies a copy until EOF
+     */
+    err = kbb_create(&kbb, kbd->cn, NULL);
+    if (err)
+        return err;
+
+    if (start) {
+        char kbuf[HSE_KVS_KEY_LEN_MAX];
+        struct kvs_ktuple seek;
+        uint klen;
+
+        key_obj_copy(kbuf, sizeof(kbuf), &klen, start);
+        kvs_ktuple_init_nohash(&seek, kbuf, klen);
+
+        err = wbti_create(&wbti, kbd->kd_mbd, kbd->kd_wbd, &seek, false, false);
+    } else {
+        err = wbti_create(&wbti, kbd->kd_mbd, kbd->kd_wbd, NULL, false, false);
+    }
+
+    while (!err && wbti_next(wbti, &cur.ko_sfx, &cur.ko_sfx_len, &kmd)) {
+        struct kbb_key_stats stats = { 0 };
+        size_t off = 0, kmd_cnt_off;
+        uint nvals;
+
+        wbti_prefix(wbti, &cur.ko_pfx, &cur.ko_pfx_len);
+
+        if (!start || key_obj_cmp(&cur, start) > 0) {
+            start = NULL; /* skip future comparisons with start */
+            if (end && key_obj_cmp(&cur, end) > 0)
+                break;
```

Comment:
I would expect the seek used with create means we don't ever need to compare against start here.

Reply:
Ahh... it's in case the key equals start, and start is exclusive.

Reply:
That extra level of nesting for what is really a loop start and stop condition throws me off.  Something like this makes it more clear to me what's going on:
```
while (!err && wbti_next(wbti, &cur.ko_sfx, &cur.ko_sfx_len, &kmd)) {

    if (HSE_UNLIKELY(start && key_obj_cmp(&cur, start) == 0))
        continue;
    start = NULL;

    if (HSE_UNLIKELY(end && key_obj_cmp(&cur, end) > 0)_
        break;

    while (nvals--) {
        ,,,,
    }
```

Reply:
I'm not a big fan of continue inside a loop. But if you think what you've suggested is more clear, I can change it.

Reply:
good point!

Reply:
Just curious, what don't you like about continue?

Reply:
I'm changing it..

Reply:
a break is no different.  just more common.   a real loop only ends when loop condition is false :)

Reply:
Agreed :-)

Reply:
I noodled with advancing past start before entering while, but then you'd have to carry the key forward if start != split.  No really clean answer here.

Reply:
Yep, I went that route initially but found this to read a bit better..